### ISSUE TITLE
real-time prediction using RNN/LSTM (different input shape for training and predicting)

### ISSUE THREAD

**Issue by pxlong** [fchollet/keras#2921]
Hey guys,

I'd like to design a RNN/LSTM network to predict in real time, i.e. once received an input at a time t, the network will give an output y_t. And the input from t=0 to t=n is a sequence. 
I have read all tutorials/examples in Keras related to RNN/LSTM, but still can not find a way to solve the problem.

Let me explain the problems with two figures:
The input at training period is illustrated in the following figure (the first one), and you can also check the second figure for input fed into the model at prediction period.
Please correct me if something is wrong in the figures.

![training_input](https://cloud.githubusercontent.com/assets/4568173/15862750/2d578e34-2d03-11e6-80d3-830fc1e7b23c.png)
![predict_input](https://cloud.githubusercontent.com/assets/4568173/15862749/2ce1c136-2d03-11e6-82c6-caea0b9456b1.png)

As illustrated in the figures, an example fed into the network is a sequence at the training period, but when I use the **trained** network to predict in real time, the input fed into the network is a data point of a sequence (i.e. features at time t of a sequence) rather than a whole sequence: 
1. so how should I structure the input to make the model works **at both training and prediction sessions**? How would my input be at the training period? 
2. Since the length of a sequence is varied, should I use marks/tokens like "START", "END" to label the start/end position of a sequence? If so, is there any examples/tutorials in Keras demonstrated it? If not, how should I process the varied sequence in Keras?

Any help will be appreciated. 
Thanks.

**Comment by sibosutd** [comment]
1. so how should I structure the input to make the model works at both training and prediction sessions? How would my input be at the training period?
A: I am guessing what you might need is stateful LSTM for training and prediction. And you have to explicitly define the batch size and time step.
https://github.com/fchollet/keras/blob/master/examples/stateful_lstm.py

2. Since the length of a sequence is varied, should I use marks/tokens like "START", "END" to label the start/end position of a sequence? If so, is there any examples/tutorials in Keras demonstrated it? If not, how should I process the varied sequence in Keras?
A: use additional dimension to denote "START" and "END" using zero or one. For sequences of varied length, padding with zeros is one possible way.

**Comment by pxlong** [comment]
@sibosutd Hi, really thanks for your reply. That's helpful.

For padding the sequences, do I need to pad the output (y_t) too?

**Comment by sibosutd** [comment]
@pxlong You don't have to pad test data with zeros. Output sequence will be terminated when sampling an "END". And padding is just a way to align your training data as matrix representation.

**Comment by pxlong** [comment]
@sibosutd Sure, I will not pad test data. 

What I plan to do is padding the variable inputs with zeros first, then using a masking layer before LSTM layers. Is that right? 
If so, do I need to pad outputs too? Since outputs are also variable just like the inputs.

**Comment by sibosutd** [comment]
@pxlong What I plan to do is padding the variable inputs with zeros first, then using a masking layer before LSTM layers. Is that right? 
A: Yes, masking can be utilized with an Embedding layer. See: **Masking** part from http://keras.io/layers/recurrent/

If so, do I need to pad outputs too? Since outputs are also variable just like the inputs.
A: Ok.. so when you say outputs, you mean shifted input as ground truth, right? If so, yes,  of course they also need to be padded.

**Comment by pxlong** [comment]
@sibosutd 

1. Sorry for unclear description, I am not doing a NLP job, so I will not use an Embedding layer. Just padding the sequences, then using a Masking layer: http://keras.io/layers/core/#masking

2. I am using a many-to-many model (give the model an input x_t (360 dims), it returns an output y_t (two dims) at each time step, so the length of the output is exactly same with the input.
Maybe I do not need to pad the output, but I think I need a way to **mask the costs** using the same variable as the input masking so only outputs for valid inputs are considered.

**Comment by sibosutd** [comment]
@pxlong 
It's ok. I am also learning. ;)

**Comment by munichong** [comment]
Hi, @sibosutd and @pxlong 

I am facing a very similar research problem. The only difference is that at a time step of a sequence, say t=1, I already know y_0. So I want to use observed x0, y0, and x1 to predict y1. Then after I observe y1, at the next time step, I want to predict y3, using x0, y0, x1, y1, and x3. 

Is it a 'many to many' or 'many to one' problem? 

I am not sure if I make my problem clear. The detail is [here](http://stats.stackexchange.com/questions/232519/how-should-i-use-recurrent-neural-network-to-model-this-problem).

Thanks in advance!

**Comment by tinaty** [comment]
@munichong do you solve your problem?
